Core/include/Acts/Propagator/Propagator.ipp

```diff
@@ -27,80 +30,154 @@ template <typename S, typename N>
-  // a positive outcome.
-
-  // start at true, if we don't begin the stepping loop we're fine.
-  bool terminatedNormally = true;
-
-  // Pre-Stepping: abort condition check
-  if (!state.options.actorList.checkAbort(state, m_stepper, m_navigator,
-                                          logger())) {
-    // Stepping loop
-    ACTS_VERBOSE("Starting stepping loop.");
-
-    terminatedNormally = false;  // priming error condition
-
-    // Propagation loop : stepping
-    for (; state.steps < state.options.maxSteps; ++state.steps) {
-      // Pre-Stepping: target setting
-      state.stage = PropagatorStage::preStep;
-      m_navigator.preStep(state, m_stepper);
-      // Perform a propagation step - it takes the propagation state
-      Result<double> res = m_stepper.step(state, m_navigator);
-      if (res.ok()) {
-        // Accumulate the path length
-        double s = *res;
-        state.pathLength += s;
-        ACTS_VERBOSE("Step with size = " << s << " performed");
-      } else {
-        ACTS_ERROR("Step failed with " << res.error() << ": "
-                                       << res.error().message());
-        // pass error to caller
-        return res.error();
+
+  if (state.options.actorList.checkAbort(state, m_stepper, m_navigator,
+                                         logger())) {
+    ACTS_VERBOSE("Propagation terminated without going into stepping loop.");
+
+    state.stage = PropagatorStage::postPropagation;
+
+    state.options.actorList.act(state, m_stepper, m_navigator, logger());
+
+    return Result<void>::success();
+  }
+
+  auto getNextTarget = [&]() -> Result<NavigationTarget> {
+    for (unsigned int i = 0; i < state.options.maxTargetSkipping; ++i) {
+      NavigationTarget nextTarget = m_navigator.estimateNextTarget(
+          state.navigation, state.position, state.direction);
+      if (!nextTarget.isValid()) {
+        return NavigationTarget::invalid();
       }
-      // release actor and aborter constrains after step was performed
-      m_stepper.releaseStepSize(state.stepping, ConstrainedStep::actor);
-      m_stepper.releaseStepSize(state.stepping, ConstrainedStep::aborter);
-      // Post-stepping:
-      // navigator post step call - actor list act - actor list check
-      state.stage = PropagatorStage::postStep;
-      m_navigator.postStep(state, m_stepper);
-      state.options.actorList.act(state, m_stepper, m_navigator, logger());
-      if (state.options.actorList.checkAbort(state, m_stepper, m_navigator,
-                                             logger())) {
-        terminatedNormally = true;
-        break;
+      IntersectionStatus preStepSurfaceStatus = m_stepper.updateSurfaceStatus(
+          state.stepping, *nextTarget.surface,
+          nextTarget.surfaceIntersectionIndex, state.options.direction,
+          nextTarget.boundaryTolerance, state.options.surfaceTolerance,
+          logger());
+      if (preStepSurfaceStatus >= Acts::IntersectionStatus::reachable) {
```

Comment:
The `>=` operator is a bit funky. I think it's sort of clear what the comparison encodes, but can't this be achieved with a `!=` as well?

Reply:
I agree I am also not a big fan of this. It relies on the fact that we order `missed < reachable < onSurface`. Here it is catching `reachable` and `onSurface`. I will replace it by two `==`